ntoskrnl/rtl/misc.c

```diff
@@ -62,6 +62,121 @@ RtlGetVersion(IN OUT PRTL_OSVERSIONINFOW lpVersionInformation)
     return STATUS_SUCCESS;
 }
 
+/**
+ * @RtlGetNtProductType
+ *
+ * Retrieves the NT type product of the operating system. This is the kernel-mode variant
+ * of this function.
+ *
+ * @param[out]  ProductType
+ *      The NT type product enumeration value returned by the call.
+ *
+ * @return
+ *      The function returns TRUE when the call successfully returned the type product of the system.
+ *      It'll return FALSE on failure otherwise. In the latter case the function will return WinNT
+ *      as the default product type.
+ *
+ * @remarks
+ *      The call expects to be called at PASSIVE_LEVEL. This is the base implementation of the function
+ *      as there's stuff still left to be investigated. For instance the function uses ExAllocatePoolWithTag
+ *      with the tag as "iPtR". In addition to that the function checks if the product type is actually valid
+ *      by checking the "ProductTypeIsValid" member of _KUSER_SHARED_DATA structure. Currently we do not
+ *      implement code that is responsible for the management of this member, yet.
+*/
+BOOLEAN
+NTAPI
+RtlGetNtProductType(OUT PNT_PRODUCT_TYPE ProductType)
+{
+    HANDLE Key;
+    ULONG ReturnedLength;
+    OBJECT_ATTRIBUTES ObjectAttributes;
+    NTSTATUS Status;
+    UCHAR Buffer[sizeof(PKEY_VALUE_PARTIAL_INFORMATION) + 256 * sizeof(WCHAR)];
+    PKEY_VALUE_PARTIAL_INFORMATION PartialInformation = (PVOID)Buffer;
+    UNICODE_STRING NtProductType;
+    static UNICODE_STRING WorkstationProduct = RTL_CONSTANT_STRING(L"WinNT");
+    static UNICODE_STRING LanManProduct = RTL_CONSTANT_STRING(L"LanmanNT");
+    static UNICODE_STRING KeyProduct = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\SYSTEM\\CurrentControlSet\\Control\\ProductOptions");
+    static UNICODE_STRING ValueNtProduct = RTL_CONSTANT_STRING(L"ProductType");
+    PAGED_CODE();
+
+    /* Initialize the object attributes to open our key */
+    InitializeObjectAttributes(&ObjectAttributes,
+                               &KeyProduct,
+                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+                               NULL,
+                               NULL);
+
+    /* Open the key */
+    Status = ZwOpenKey(&Key, KEY_QUERY_VALUE, &ObjectAttributes);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlGetNtProductType(): The ZwOpenKey() function has failed! (Status: 0x%lx)\n", Status);
+        *ProductType = NtProductWinNt;
+        return FALSE;
+    }
+
+    /* Now it's time to query the value from the key to check what kind of product the OS is */
+    Status = ZwQueryValueKey(Key,
+                             &ValueNtProduct,
+                             KeyValuePartialInformation,
+                             PartialInformation,
+                             sizeof(Buffer),
+                             &ReturnedLength);
+    if (!NT_SUCCESS(Status))
```

Comment:
If you close the key before doing this check, you can remove the `ZwClose` from all branches below